addons/core/addon/decorators/resource-filter.js

```diff
@@ -0,0 +1,251 @@
+            this.deserializeValue(item, serializedValue)
+          )
+        )
+      : null;
+    return deserializedValue;
+  }
+  set value(value) {
+    const queryParams = {};
+    const serialized = value.map((value) => this.serializeValue(value));
+    queryParams[this.filterKey] = JSON.stringify(serialized);
+    this.route.transitionTo({ queryParams });
+  }
+
+  // =methods
+
+  /**
+   * Calls `serialize` (if set) on the passed value.  Otherwise returns the
+   * value unaltered.
+   */
+  serializeValue(value) {
+    if (this.serialize) {
+      return this.serialize(...arguments);
+    }
+    return value;
+  }
+
+  /**
+   * Tests if `item` is represented by `serializedValue` when deserializing
+   * items from the query string.  If `findBySerialized` is unspecified,
+   * a simple equality test is used.
+   * @param item
+   * @param {string} serializedValue
+   * @return {boolean} item is represented by `serializedValue`
+   */
+  deserializeValue(item, serializedValue) {
+    if (this.findBySerialized) {
+      return this.findBySerialized(...arguments);
+    }
+    return item === serializedValue;
+  }
+
+  // =static methods
+
+  /**
+   * Configures the specified route with a query parameter of `name` that, when
+   * changed, refreshes the route model and replaces history (rather than
+   * advancing history).
+   * @static
+   * @param {Route} routeClass
+   * @param {string} name
+   */
+  static setupRouteQueryParams(routeClass, name) {
+    const filterKey = `filter-${name}`;
+
+    // If the route has no query params specified yet, declare them.
+    if (!Object.keys(routeClass.queryParams).length)
+      routeClass.queryParams = {};
+
+    // If the route has no resource filters list specified yet, create one.
+    if (!routeClass.resourceFilterParams) routeClass.resourceFilterParams = [];
+
+    // Add the resource filter query parameter to the route.  Resource filter
+    // param changes refresh the model but do not contribute to browser history.
+    // See https://guides.emberjs.com/release/routing/query-params/
+    routeClass.queryParams[filterKey] = {
+      refreshModel: true,
+      replace: true,
+    };
+  }
+
+  /**
+   * Factory method to instantiate an instance of this class.
+   * @static
+   * @param {Route} routeInstance
+   * @param {string} name
+   * @param {array|function} allowed
+   * @param {?array} defaultValue
+   * @param {?function} serialize
+   * @param {?function} findBySerialized
+   * @return {RouteResourceFilter}
+   */
+  static getOrCreateRouteResourceFilter(
+    routeInstance,
+    name,
+    allowed,
+    defaultValue,
+    serialize,
+    findBySerialized
+  ) {
+    const owner = getOwner(routeInstance);
+    const containerKey = `resource-filter:${name}@${routeInstance.routeName}`;
+    const factory = owner.factoryFor(containerKey);
+    if (!factory) {
+      // create factory
+      class AppliedRouteResourceFilter extends this {
```

Comment:
Is this class declaration extending from `RouteResourceFilter` ?

Reply:
Yes.  `this` in the context refers to the class itself, because the method is `static`.